cpp/src/arrow/dataset/discovery.h

```diff
@@ -59,21 +61,13 @@ class ARROW_DS_EXPORT DataSourceDiscovery {
   /// \brief Get unified schema for the resulting DataSource.
   virtual Result<std::shared_ptr<Schema>> Inspect();
 
-  /// \brief Create a DataSource with a given partition.
-  virtual Result<DataSourcePtr> Finish() = 0;
+  /// \brief Create a DataSource with the given schema.
+  virtual Result<DataSourcePtr> Finish(const std::shared_ptr<Schema>& schema) = 0;
 
-  std::shared_ptr<Schema> schema() const { return schema_; }
-  Status SetSchema(std::shared_ptr<Schema> schema) {
-    schema_ = schema;
-    return Status::OK();
-  }
-
-  const PartitionSchemePtr& partition_scheme() const { return partition_scheme_; }
-  Status SetPartitionScheme(PartitionSchemePtr partition_scheme) {
-    partition_scheme_ = partition_scheme;
-    return Status::OK();
-  }
+  /// \brief Create a DataSource using an inspected schema.
+  Result<DataSourcePtr> Finish();
```

Comment:
Use an optional parameter and merge the previous definitions. See `discovery.cc` about the behavior of this function, e.g. validation of explicit schema.

Reply:
alright

Reply:
Actually I'm not sure I agree with this one; default parameters in a virtual method pass the default case handling down to each implementation. Is there any behavior which should be supported for "Finish without explicit schema" other than "inspect a schema then finish with that"?

Reply:
Then let's remove the virtual case for now :)

Reply:
I think it's an anti-pattern to expect the user to call the Validate method so it doesn't blow up in his face. Anyhow, I'll tackle this in my next PR on DatasetDiscovery as discrepency will likely happen there more than in the DataSourceDiscovery.